### ISSUE TITLE
EXTREMELY SLOW (unusable) towards end of tokenization of dataset with long multi turn conversations

### ISSUE THREAD

**Issue by Nero10578** [axolotl-ai-cloud/axolotl#2396]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Expected tokenization to work extremely fast as in the old commits in from this commit and older 339f3c67e2d6855340b5958274ea539517829baa

### Current behaviour

Ever since a change to how the chat_templates works back in this commit 10cfecf02e8829de749708c2588dc76be3a156d2 the tokenization has become literally unusable for my dataset that contains long multi turn conversation. The speed seems to become exponentially slower towards the end of the tokenization since it uses less and less CPU cores as it finishes most of the examples.

Since that commit I have just been replacing the chat_template.py with the older one and it worked fine, but now since this commit b194e17c2847348bb102d24e86775dfb77c9c5fe this hacky fix stopped working as the new chat_template.py is much different. So now newer axolotl version are unusable to me.

![Image](https://github.com/user-attachments/assets/faeec121-7aa7-41cd-8494-068217a94c4f)

I can help the performance to be better by changing num_proc in datasets.py to a huge number but it still eventually slows to a crawl at the end.

### Steps to reproduce

Run preprocess on using the latest axolotl commit on a long multi turn conversation dataset.

### Config yaml

```yaml
base_model: /home/owen/models/Mistral-Small-24B-Instruct-2501
tokenizer_type: AutoTokenizer

train_on_inputs: false
group_by_length: false
load_in_8bit: false
load_in_4bit: true
strict: false

sequence_len: 8192
bf16: auto
tf32: false
flash_attention: true

shuffle_merged_datasets: true

# lora_on_cpu: true
# gpu_memory_limit: 20GiB

#Data
datasets:
  - path: /home/owen/datasets/RPMax-v1.4-formatted/conversations-escaped-formatted_split_1.json
    type: chat_template
    field_messages: conversations
    message_field_role: from
    message_field_content: value


warmup_steps: 20
dataset_prepared_path: ./lora_last_run_prepared

# Iterations
num_epochs: 1
saves_per_epoch: 8
saves_total_limit: 8

# Evaluation
val_set_size: 0.0025
eval_max_new_tokens: 128
eval_sample_packing: false
evals_per_epoch: 8

# LoRA
output_dir: ./qlora_out_rex
adapter: qlora
lora_model_dir:
lora_r: 128
lora_alpha: 128
lora_target_linear: true

save_safetensors: true

peft_use_rslora: false
loraplus_lr_ratio: 16

# Sampling
sample_packing: true
pad_to_sequence_len: true

# Batching
gradient_accumulation_steps: 16
micro_batch_size: 1
gradient_checkpointing: false
gradient_checkpointing_kwargs:
  use_reentrant: false

# wandb
wandb_mode: # "offline" to save run metadata locally and not sync to the server, "disabled" to turn off wandb
wandb_project: 
wandb_entity: # A wandb Team name if using a Team
wandb_watch:
wandb_name:
wandb_run_id: # Set the ID of your wandb run
wandb_log_model: # "checkpoint" to log model to wandb Artifacts every `save_steps` or "end" to log only at the end of training

# Optimizer
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 0.00001

# Misc
auto_resume_from_checkpoints: true
logging_steps: 1
weight_decay: 0.0

special_tokens:
  pad_token: <pad>

plugins:
  - axolotl.integrations.liger.LigerPlugin
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
cut_cross_entropy: true
liger_rope: true
liger_rms_norm: true
liger_layer_norm: true
liger_glu_activation: true
#liger_fused_linear_cross_entropy: true

# Multi-GPU
# deepspeed: /home/owen/axolotl/deepspeed_configs/zero3_bf16_cpuoffload_params.json
fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: false
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: MistralDecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_activation_checkpointing: true
```

### Possible solution

I think what chat_template.py does is find start and end points for each turn by checking the raw tokens? Shouldn't it just use the json structure instead when it can, where there is already a clear conversation start and end point for each turn.

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main/16dc6ee68d05bfbef607a8ee06e00fca1d8c4739

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hello! Thanks for reporting this. I was informed about this as well in Discord (maybe by you).

It makes sense that it would take longer for extended multi-turn convo as we go through turn by turn in this processing to be able to detect start and end boundaries of a turn (we use a dummy turn with the actual to compare where it differs). I'm very open to improving this.

> I think what chat_template.py does is find start and end points for each turn by checking the raw tokens? Shouldn't it just use the json structure instead when it can, where there is already a clear conversation start and end point for each turn.

Could you elaborate what this means?

**Comment by Nero10578** [comment]
> Hello! Thanks for reporting this. I was informed about this as well in Discord (maybe by you).
> 

Not sure if I messaged you in discord, but I did mention this to some other people in a few discords I am in.

> It makes sense that it would take longer for extended multi-turn convo as we go through turn by turn in this processing to be able to detect start and end boundaries of a turn (we use a dummy turn with the actual to compare where it differs). I'm very open to improving this.
> 

Is it possible to just not queue a lot of examples per thread from the beginning? The bottleneck is having a large amount of long examples per thread being statically queued from the beginning. If we keep distributing the load to separate cores all the way until all the examples are processed it would make better use of all the CPU cores.

> > I think what chat_template.py does is find start and end points for each turn by checking the raw tokens? Shouldn't it just use the json structure instead when it can, where there is already a clear conversation start and end point for each turn.
> 
> Could you elaborate what this means?

As far as I understand the code, it checks for the start and end boundaries via matching the tokens am I right? Is it not possible to just take the conversations structure from the data JSON file? It already is separated into user and assistant messages anyways in the JSON file if you use the conversations format no?

**Comment by NanoCode012** [comment]
> Is it possible to just not queue a lot of examples per thread from the beginning?

We don't actually manage the queue ourselves. We just take advantage of `datasets.map` and provide the function that runs to it https://huggingface.co/docs/datasets/v3.3.2/en/package_reference/main_classes#datasets.Dataset.map

>  it checks for the start and end boundaries via matching the tokens am I right? Is it not possible to just take the conversations structure from the data JSON file?

Yes, it is already separated, however, this is done to check for any delimiter tokens.

This is how it's currently done:
```
<user>Real convo<eot>

<user>Fake message<eot>
```
We then use the above to detect the start (ind 6) and end (ind -5) difference. This is then done per turn (as turns get longer, the fake message is added at the last index of the turns). 
```
<user>Real convo<eot><model>Real convo<eot>

<user>Real convo<eot><model>Fake message<eot>
```

The reason we call `apply_chat_template` is because with the various chat_templates available, we can't really generalize how each would look. Some has EOT, some doesn't. Some use different/hardcoded message that only appear when in certain roles e.g. `tools`. I also think it's not possible to only process the last turn each time because some templates require user/assistant/user alternating turns or others.

If there's a much more optimized way, I'm very open to this change.

**Comment by Nero10578** [comment]
> > Is it possible to just not queue a lot of examples per thread from the beginning?
> 
> We don't actually manage the queue ourselves. We just take advantage of `datasets.map` and provide the function that runs to it https://huggingface.co/docs/datasets/v3.3.2/en/package_reference/main_classes#datasets.Dataset.map

A workaround of sorts to this is that I have been increasing the num_proc to a ridiculous number and that already helps a lot as it reduces the number of examples stuck to a single thread queue. Not ideal though and still goes extremely slowly towards the end.

> 
> > it checks for the start and end boundaries via matching the tokens am I right? Is it not possible to just take the conversations structure from the data JSON file?
> 
> Yes, it is already separated, however, this is done to check for any delimiter tokens.
> 
> This is how it's currently done:
> 
> ```
> <user>Real convo<eot>
> 
> <user>Fake message<eot>
> ```
> 
> We then use the above to detect the start (ind 6) and end (ind -5) difference. This is then done per turn (as turns get longer, the fake message is added at the last index of the turns).
> 
> ```
> <user>Real convo<eot><model>Real convo<eot>
> 
> <user>Real convo<eot><model>Fake message<eot>
> ```
> 
> The reason we call `apply_chat_template` is because with the various chat_templates available, we can't really generalize how each would look. Some has EOT, some doesn't. Some use different/hardcoded message that only appear when in certain roles e.g. `tools`. I also think it's not possible to only process the last turn each time because some templates require user/assistant/user alternating turns or others.
> 
> If there's a much more optimized way, I'm very open to this change.

I see. So the change is to support the wider range of chat templates? Maybe a config option to use the old method from https://github.com/axolotl-ai-cloud/axolotl/commit/339f3c67e2d6855340b5958274ea539517829baa instead is the best way then? This is just unusable on long multi turns at the moment, I can leave it for hours and it won't complete the last few hundred examples since it starts taking multiple seconds per example.

**Comment by NanoCode012** [comment]
I can't remember off the top of my head, but the old method had issues with some tokenizers (Mistral) causing masking to be off, which would have hidden training issues (please check your debug outputs to ensure it's correct). I would really not recommend using it.

>  I can leave it for hours and it won't complete the last few hundred examples since it starts taking multiple seconds per example.

This may be more of an inherent issue with the `.map` right? I'm actually curious why it starts slow -> fast midway -> slow near the end. Would it be better to ask in the HF datasets repo on why that is setup like that?

or, if there's some way to cache some of the operations during the `get_turn` , that would also help. I would assume that to be the main time offender.

**Comment by NanoCode012** [comment]
@Nero10578 , do you know whether you can run the preprocessing on a separate CPU-only machine first, upload to HF via `push_dataset_to_hub`, then use it for your GPU runs? This should alleviate your long wait times

**Comment by Nero10578** [comment]
Sorry for the late reply on this.

> I can't remember off the top of my head, but the old method had issues with some tokenizers (Mistral) causing masking to be off, which would have hidden training issues (please check your debug outputs to ensure it's correct). I would really not recommend using it.
> 

I've checked the debug when I ran this mistral small and it looks correct to me, but yes its possible I guess some of the examples have it be masked wrongly. It does seem to train just fine on my dataset, and at the moment I am forced to use the old chat_template method since the new one just literally will never finish tokenizing. The iteration speeds seems to keep getting slower and slower to the point it virtually never finishes.

> > I can leave it for hours and it won't complete the last few hundred examples since it starts taking multiple seconds per example.
> 
> This may be more of an inherent issue with the `.map` right? I'm actually curious why it starts slow -> fast midway -> slow near the end. Would it be better to ask in the HF datasets repo on why that is setup like that?
> 
> or, if there's some way to cache some of the operations during the `get_turn` , that would also help. I would assume that to be the main time offender.

To be honest I am not sure where the root of the issue is, all I know is it started being broken when the chat_template masking got changed to the more complicated new method. It makes preprocessing the dataset slower across any dataset I have and the dataset with long multi turns are basically impossible to finish preprocessing.

> [@Nero10578](https://github.com/Nero10578) , do you know whether you can run the preprocessing on a separate CPU-only machine first, upload to HF via `push_dataset_to_hub`, then use it for your GPU runs? This should alleviate your long wait times

I am already running the preprocess on the fastest CPU I own which is in my training machine. Its an AMD Threadripper 3960X that runs at 4.3GHz and I can see as the number of processes goes down as they complete the CPU utilization goes down until its basically only 1-3 cores left running at 100% load while the rest of the CPU is idle. I think this issue can be alleviated if we can somehow redistribute the examples to more threads towards the end of the preprocessing.

**Comment by NanoCode012** [comment]
Thanks for the follow up. We had an internal discussion about this. We currently suspect the numerous calls to tokenizer to be the bottleneck, so I will do a bit of benchmarking to check where the slowdown is and perhaps batch call to the tokenizer instead of individually call like we're doing now.

For your dataset, could you share a bit about it? How many turns are there in one sample? How many tokens are there in average in one sample? How many samples are there?

**Comment by tnlin** [comment]
Similar issue here, any progress? My dataset is about  1M samples,  with an average of 5k input / 50 output tokens.  (average 30~40 turn)

**Comment by maziyarpanahi** [comment]
to also update, i have faced a similar issue. the tokenization of a large dataset started with 3k a second, but as you can see at the end we are at 30/s throughput! 

<img width="413" height="467" alt="Image" src="https://github.com/user-attachments/assets/b7725e9c-a142-46f8-8600-9c3681692775" />

**Comment by NanoCode012** [comment]
Hey, I think we found a shuffle helped with distributing long sequences to different processes. We fixed this a while back. Are you on our latest release?

**Comment by maziyarpanahi** [comment]
> Hey, I think we found a shuffle helped with distributing long sequences to different processes. We fixed this a while back. Are you on our latest release?

hi @NanoCode012 

yes, i am using the latest docker image with "git fetch/pull" to be on the main branch. are you talking about `shuffle_merged_datasets` param in the config?

**Comment by NanoCode012** [comment]
Yep. It should be activated by default. How long are your sequences? Are they all the same lengths or varied?

**Comment by maziyarpanahi** [comment]
i have tested `shuffle_merged_datasets` being set to True. still each dataset becomes slow (10 examples/s) towards the end after 200k or so. 

so if you have 4 or 5 datasets all above 200k, it will take hours just because of what happens at the end of each one

**Comment by ErenAta16** [comment]
I read through the current `chat_template.py` and the dataset-loading pipeline to pin down both halves of this (why long multi-turn conversations are disproportionately expensive, and why they specifically show up as "slows down and uses fewer cores near the end"). Both are confirmable from the code as it stands today, no speculation needed.

**Part 1: `find_turn` is O(turns^2) per conversation, not O(turns)**

`ChatTemplateStrategy._tokenize_single_prompt` (chat_template.py, the training-detail path starting ~line 477) calls `self.find_turn(...)` once for every trainable turn. Look at what `find_turn` does per call (chat_template.py:663-736):

```python
turns_with_empty = turns[:turn_idx] + [empty_turn]
turns_with_content = turns[: turn_idx + 1]
...
dummy_ids = _extract_input_ids(self.prompter.build_prompt(turns_with_empty, ...))
full_ids = _extract_input_ids(self.prompter.build_prompt(turns_with_content, ...))
```

`build_prompt` (chat_template.py:103-163) calls `self.tokenizer.apply_chat_template(conversation, tokenize=True, ...)` directly. So for turn `k`, this re-renders the Jinja template and re-tokenizes the entire conversation prefix `turns[0..k]` from scratch, twice. It does this for every turn from 0 to T. Total tokenizer/template work across one conversation is `sum_{k=0}^{T} O(k)` = O(T^2), not O(T) or O(total token count). A 40-turn conversation (as reported by @tnlin) doesn't do ~10x the retokenization work of a 4-turn one, it does closer to (40/4)^2 = 100x, because every one of those 40 turns re-renders and re-tokenizes everything before it again.

This is exactly @NanoCode012's suspicion ("numerous calls to tokenizer") but sharper: it's not just call count, each call's cost also grows with how far into the conversation you are, which is what produces the specific "fine at first, catastrophic on long conversations" shape everyone is reporting.

**Part 2: the shipped mitigation (`shuffle_merged_datasets`) can't help this, because of pipeline ordering**

Tokenization of each dataset in your `datasets:` list happens per-source-file, in `TokenizedPromptDataset.process()` (`axolotl/datasets.py`, `dataset.map(self.prompt_tokenizer.tokenize_prompt, num_proc=self.process_count, ...)`), invoked once per entry inside the loop in `_load_raw_datasets` (`utils/data/sft.py:329-339`). `datasets.map(num_proc=N)` splits the dataset into `N` static, contiguous shards up front and does not work-steal or rebalance. Since per-example cost is quadratic in turn count (Part 1), if the long-conversation examples aren't evenly spread across the file, whatever shard(s) happen to contain them finish last while the other N-1 processes sit idle. That is the "less and less CPU cores as it finishes" symptom.

`merge_datasets` (`utils/data/shared.py:685-720`), where `shuffle_merged_datasets` actually shuffles, is only called at `sft.py:344`, **after** the per-dataset tokenization loop above has already run to completion. Shuffling the merged, already-tokenized dataset afterward cannot retroactively rebalance work that has already been done on the un-shuffled shards. This lines up exactly with what @maziyarpanahi reported after enabling it: "still each dataset becomes slow... towards the end" (Sept 2025 comment). The shuffle isn't ineffective because of variance surviving randomization, it's ineffective because it runs one pipeline stage too late to matter for this specific bottleneck.

**Fix directions**

1. Actual root cause (Part 1): stop re-deriving the whole prefix from scratch at every turn. Since `_tokenize_single_prompt` already walks turns in increasing order, `find_turn` could carry forward the `full_ids` computed at turn `k-1` as the base for turn `k`'s probe, and only tokenize the newly appended turn's delta, falling back to a full from-scratch render only where template behavior is position-sensitive near the true end of the conversation (the code already special-cases this via `real_last_index`, e.g. the Mistral "no output for lone system message" branch at chat_template.py:685-689). That takes the per-conversation cost from O(T^2) to roughly O(T), independent of any dataset-level scheduling.
2. Mitigate the imbalance regardless (Part 2): shuffle (or at least round-robin) each individual dataset's row order *before* it goes into `TokenizedPromptDataset.process()`'s `.map(num_proc=...)` call in `_load_and_process_single_dataset`, not only in `merge_datasets` afterward. `shared.py` already has a structurally similar `shuffle_before_merging_datasets` flag; the same idea just needs to run one stage earlier, before tokenization rather than after it.

These are complementary: #1 removes the wasted work entirely, #2 makes sure whatever work remains is spread evenly across `num_proc` workers. Doing only #2 (which is what shipped) can't fix this given the current call order, and is consistent with every report in this thread post-dating that change still showing the tail-end slowdown.
